doc: add inheritance and composition section in user docs (#2165)
copier-org/copier: docs/inheritance_and_composition.md

Author: My feeling is that `reusing` is too generic, and it is especially confusing in the context of copier. The first thing that would come to my mind is "Of course! I have some code that I want to re-use by creating a template! Let's see how".

What about `composing`, `composing_templates`, or `inheriting_and_composing_templates` (too long perhaps)? Alternatively, `reusing_templates` may be explicit enough.

I leave the final decision to you / the maintainers, just let me know what do you think is best and I will apply 😄

# Template Inheritance and Composition

> ## What is Inheritance, what is Composition?

Reviewer: Perhaps this is more concise?

```suggestion
## Inheritance vs. Composition
```

> Template Inheritance is a way to create a new template based on an existing one. It

Reviewer: * I'd not capitalize "Inheritance" in a regular sentence.
* How about introducing "parent" and "child" in the first sentence to clarify the relationship for later use?

```suggestion
Template inheritance is a way to create a new (child) template based on an existing (parent) template. It
```

allows you to reuse code and structure from the parent template while adding or

> overriding specific parts in the child template. This is useful for creating variations

Reviewer: I think it's mostly about _extensions_, not so much _variations_ – perhaps a matter of definition. Variations – in my mind – can often be addressed by conditional questions and conditional rendering in the same template. WDYT?

```suggestion
overriding specific parts in the child template. This is useful for creating extensions
```

of a template without duplicating code.

<!-- TODO: add an example of inheritance -->

> Template Composition, on the other hand, is a way to combine multiple templates,

Reviewer: Like above:

```suggestion
Template composition, on the other hand, is a way to combine multiple templates,
```

independent of each other, into a single new template. In this way you can join the

> functionalities of multiple templates, by adding some "glue code", and avoid repetition.

Reviewer: Do you have an approach like yours using tasks that run `copier copy` in mind? I've mostly thought of "composition" along the lines of https://copier.readthedocs.io/en/stable/configuring/#applying-multiple-templates-to-the-same-subproject. There, template composition means generating output from multiple templates in the same target directory, but there is no new template that is a composition of other templates.

<!-- TODO: add an example of composition -->

<!-- TODO? (not sure if it relevant...) add an operational distinction between the two -->

## Why copier does not natively support inheritance and composition?

> <!-- TODO: add content, re-use content from https://github.com/copier-org/copier/issues/934#issuecomment-1518964035 -->

Reviewer: I anticipate that a proper solution might require Copier to support it natively. At least for the fork-like workflow, I think that specified template versions in parent template migrations need to be mapped to child template versions – at the very least. But it's too early to make a definitive statement.

Author: Reading carefully what he wrote once again, it seems to me that everything reduces to the UNIX philosophy of "do one thing well", or am I wrong?

## Solutions

> ### Git submodules + jinja template inheritance

Reviewer: ```suggestion
### Git submodules + Jinja template inheritance
```

:nerd_face:

<!-- TODO: Add content-->

### Fork-like approach

<!-- TODO: Add content-->

> ### Leverage on copier task

Reviewer: Perhaps something like this?

```suggestion
### Delegated `copier copy` calls via tasks
```

Not sure ... :thinking:
